File: RotateList.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode rotateRight(ListNode head, int k) {
        if (k <= 0 || head == null || head.next == null) {
            return head;
        }

        ListNode last = head;
        int length = 1;
        while (last.next != null) {
            last = last.next;
            length++;
        }

        last.next = head;
        int rotations = k % length;
        int skip = length - rotations;
        ListNode newLast = head;
        for (int i = 0; i < skip - 1; i++) {
            newLast = newLast.next;
        }
        head = newLast.next;
        newLast.next = null;

        return head;
    }
}

// 2

class Solution {
    public ListNode rotateRight(ListNode head, int k) {
       if(head == null || head.next == null){
            return head;
        }
        ListNode dummy = new ListNode(0);
        dummy.next = head;
        ListNode fast = dummy;
        ListNode slow = dummy;
        int i;
        for(i = 0; fast.next != null; i++){
            fast = fast.next;
        }
        for(int j = i - k % i; j > 0; j--){
            slow = slow.next;
        }
        fast.next = dummy.next;
        dummy.next = slow.next;
        slow.next = null;
        return dummy.next; 
    }
}

File: RotateList.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* rotateRight(ListNode* head, int k) {
        if(head == NULL && head->next == NULL){
            return head;
        }
        
        ListNode* ans = new ListNode(0);
        ans->next = head;
        ListNode* fast = ans;
        ListNode* slow = ans;
        
        int i;
        for(i=0;fast->next!=NULL;i++){
            fast = fast->next;
        }
        
        for(int j=i-k%i;j>0;j--){
            slow = slow->next;
        }
        
        fast->next = ans->next;
        ans->next = slow->next;
        slow->next = NULL;
        return ans->next;
    }
};
